drivers/interrupt_controller/intc_gpio_stm32.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (c) 2016 Open-RnD Sp. z o.o.
+ * Copyright (c) 2017 RnDity Sp. z o.o.
+ * Copyright (c) 2019-23 Linaro Limited
+ * Copyright (c) 2025 Alexander Kozhinov <ak.alexander.kozhinov@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Driver for External interrupt/event controller in STM32 MCUs
+ */
+
+#include <zephyr/device.h>
+#include <soc.h>
+#include <stm32_ll_bus.h>
+#include <stm32_ll_gpio.h> /* For STM32F1 series */
+#include <stm32_ll_exti.h>
+#include <stm32_ll_system.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/dt-bindings/pinctrl/stm32-pinctrl-common.h> /* For STM32L0 series */
+#include <zephyr/drivers/interrupt_controller/gpio_intc_stm32.h>
+#include <zephyr/drivers/interrupt_controller/intc_exti_stm32.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/irq.h>
+
+#include "stm32_hsem.h"
+#include "intc_exti_stm32_priv.h"
+
+/** @brief EXTI lines range mapped to a single interrupt line */
+struct stm32_exti_range {
+	/** Start of the range */
+	uint8_t start;
+	/** Range length */
+	uint8_t len;
+};
+
+#define EXTI_NUM_LINES_TOTAL	DT_PROP(DT_NODELABEL(exti), num_lines)
+#define NUM_EXTI_LINES			DT_PROP(DT_NODELABEL(exti), num_gpio_lines)
+
+BUILD_ASSERT(EXTI_NUM_LINES_TOTAL >= NUM_EXTI_LINES,
```

Comment:
@mathieuchopstm Unfortunately github's tracking system starts getting a mess 😄 I've removed a test case in intc_exti test folder, where `EXTI_NUM_LINES_TOTAL >= NUM_EXTI_LINES` assert test case were executed and all your comments are lost.
Anyways - I see your idea behind preprocessor based build assured and I've implemented it right here as you requested. Also assert condition is now changed to cover cases, where total number of lines is same as the number of GPIO lines. The corresponding test case were removed.